src/ReusableRelayBuilder.php

```diff
@@ -0,0 +1,39 @@
+<?php
+namespace Relay;
+
+use ArrayObject;
+use InvalidArgumentException;
+
+class ReusableRelayBuilder
+{
+    protected $resolver;
+
+    public function __construct($resolver = null)
+    {
+        $this->resolver = $resolver;
+    }
+
+    public function newInstance($queue)
+    {
+        return new ReusableRelay(new RelayFactory(
+            $this->getArray($queue),
+            $this->resolver
+        ));
+    }
+
+    protected function getArray($queue)
+    {
+        if (is_array($queue)) {
+            return $queue;
+        }
+
+        $getArrayCopy = $queue instanceof GetArrayCopyInterface
+            || $queue instanceof ArrayObject;
+
+        if ($getArrayCopy) {
+            return $queue->getArrayCopy();
```

Comment:
Would be more useful to use `iterator_to_array` instead of array copying? The code could look like this:

```php
protected function getArray($queue)
{
    if (is_array($queue)) {
        return $queue;
    }
    return iterator_to_array($queue);
}
```

This way you can bypass the `instanceof` checks entirely and rely on builtins.